extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -158,35 +160,42 @@ void prepareBouncyCastleProviders(BuildProducer<ReflectiveClassBuildItem> reflec
             runtimeReInitialized
                     .produce(new RuntimeReinitializedClassBuildItem(
                             "org.bouncycastle.jsse.provider.DefaultSSLContextSpi$LazyManagers"));
-            prepareBouncyCastleProvider(reflection, runtimeReInitialized, bouncyCastleJsseProvider.get().isInFipsMode());
+            prepareBouncyCastleProvider(curateOutcomeBuildItem, reflection, runtimeReInitialized,
+                    bouncyCastleJsseProvider.get().isInFipsMode());
         } else {
             Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider = getOne(bouncyCastleProviders);
             if (bouncyCastleProvider.isPresent()) {
-                prepareBouncyCastleProvider(reflection, runtimeReInitialized, bouncyCastleProvider.get().isInFipsMode());
+                prepareBouncyCastleProvider(curateOutcomeBuildItem, reflection, runtimeReInitialized,
+                        bouncyCastleProvider.get().isInFipsMode());
             }
         }
     }
 
-    private static void prepareBouncyCastleProvider(BuildProducer<ReflectiveClassBuildItem> reflection,
-            BuildProducer<RuntimeReinitializedClassBuildItem> runtimeReInitialized,
-            boolean isFipsMode) {
+    private static void prepareBouncyCastleProvider(CurateOutcomeBuildItem curateOutcomeBuildItem,
+            BuildProducer<ReflectiveClassBuildItem> reflection,
+            BuildProducer<RuntimeReinitializedClassBuildItem> runtimeReInitialized, boolean isFipsMode) {
         reflection.produce(new ReflectiveClassBuildItem(true, true,
                 isFipsMode ? SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_CLASS_NAME
                         : SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME));
-        reflection.produce(new ReflectiveClassBuildItem(true, true,
-                "org.bouncycastle.jcajce.provider.symmetric.AES",
-                "org.bouncycastle.jcajce.provider.symmetric.AES$CBC",
-                "org.bouncycastle.crypto.paddings.PKCS7Padding",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyFactorySpi",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyFactorySpi$EC",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyFactorySpi$ECDSA",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyPairGeneratorSpi",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyPairGeneratorSpi$EC",
-                "org.bouncycastle.jcajce.provider.asymmetric.ec.KeyPairGeneratorSpi$ECDSA",
-                "org.bouncycastle.jcajce.provider.asymmetric.rsa.KeyFactorySpi",
-                "org.bouncycastle.jcajce.provider.asymmetric.rsa.KeyPairGeneratorSpi",
-                "org.bouncycastle.jcajce.provider.asymmetric.rsa.PSSSignatureSpi",
-                "org.bouncycastle.jcajce.provider.asymmetric.rsa.PSSSignatureSpi$SHA256withRSA"));
+
+        if (curateOutcomeBuildItem.getApplicationModel().getDependencies().stream().anyMatch(
```

Comment:
```suggestion
        if (curateOutcomeBuildItem.getApplicationModel().getDependencies().stream().filter(Dependency::isRuntimeCP).anyMatch(
```
If you want to limit the check only to the runtime classpath.

Reply:
That might be a good addition. This way we could avoid registering classes that are in the classpath during native-image compilation (and thus would get registered without any warning) but not at runtime, which means that they should not get reflectively accessed at runtime anyway. I wonder how common this case might be though :/